### PR TITLE
[Fix] OR statements not being evaluated as part of nested line filters

### PR DESCRIPTION
**Pull request by poyzannur**
**What this PR does / why we need it**:
When nester line filters were being evaluated, the `OR` statements in right expressions were omitted and only `LineFilters` were returned. Resulting in only first value being returned.
It can be reproduced in stringer unit test failing with, 
```
Error Trace:	/Users/poyzannur/workspace/loki/pkg/logql/syntax/ast_test.go:535
        	Error:      	Not equal: 
        	            	expected: "{app=\"foo\"} |= \"foo\" or \"bar\" |= \"baz\" or \"bal\""
        	            	actual  : "{app=\"foo\"} |= \"foo\" or \"bar\" |= \"baz\""
```

We now return the `OR` expression as part of nested line filters. Thanks a million to @ashwanthgoli for help with debugging, and extra unit test. 

**Which issue(s) this PR fixes**:
Fixes https://github.com/grafana/support-escalations/issues/9042

**Special notes for your reviewer**:

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [x] Tests updated
- [ ] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: pkg/logql/syntax/ast_test.go [hunk 1]
```diff
@@ -383,6 +383,13 @@ func Test_FilterMatcher(t *testing.T) {
 			},
 			[]linecheck{{"foo", true}, {"bar", true}, {"none", false}},
 		},
+		{
+			`{app="foo"} |= "foo" or "bar" |= "buzz" or "fizz"`,
+			[]*labels.Matcher{
+				mustNewMatcher(labels.MatchEqual, "app", "foo"),
+			},
+			[]linecheck{{"foo buzz", true}, {"bar fizz", true}, {"bar", false}, {"none", false}},
```
Comment (chaudum) [inline pkg/logql/syntax/ast_test.go:391]:
```suggestion
			[]linecheck{{"foo buzz", true}, {"bar fizz", true}, {"foo", false}, {"bar", false}, {"none", false}},
```

File: pkg/logql/syntax/parser_test.go [hunk 2]
```diff
@@ -3138,6 +3138,29 @@ var ParseTestCases = []struct {
 			},
 		},
 	},
+	{
+		in: `{app="foo"} |= "foo" or "bar" |= "buzz" or "fizz"`,
+		exp: newPipelineExpr(
+			newMatcherExpr([]*labels.Matcher{mustNewMatcher(labels.MatchEqual, "app", "foo")}),
+			MultiStageExpr{
+				&LineFilterExpr{
+					Left: newOrLineFilter(newLineFilterExpr(labels.MatchEqual, "", "foo"), newLineFilterExpr(labels.MatchEqual, "", "bar")),
```
Comment (chaudum) [inline pkg/logql/syntax/parser_test.go:3147]: Can we construct the object in the test instead of relying on the constructor?

**Review by chaudum — COMMENTED** [review]
Please also add a changelog entry. I guess it needs to be backported as well.

**Comment by poyzannur** [comment]
> Please also add a changelog entry. I guess it needs to be backported as well.

I checked that OR statements were not released as part of 2.9, so not including a changelog entry.

**Review by chaudum — APPROVED** [review]

**Pull request merged**